src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go

```diff
@@ -100,9 +100,18 @@ func (d *SyncServerDatabase) WriteEvent(
 		}
 		streamPos = types.StreamPosition(pos)
 
-		if len(addStateEvents) == 0 && len(removeStateEventIDs) == 0 {
-			// Nothing to do, the event may have just been a message event.
-			return nil
+		fmt.Println(len(addStateEvents))
+
+		if len(addStateEvents) == 0 {
+			// If the event is a m.room.member event, and has unsigned content,
+			// we need to save it as it is very likely to be a membership update
+			// (e.g. if a user updates their profile)
+			if ev.Type() == "m.room.member" && len(ev.Unsigned()) > 0 {
+				addStateEvents = append(addStateEvents, *ev)
+			} else if len(removeStateEventIDs) == 0 {
+				// Nothing to do, the event may have just been a message event.
+				return nil
+			}
 		}
```

Comment:
We need to save the event in DB (so we can reference it in future updates) as an update of the user's membership in the room. However, the event isn't marked as being part of the current state. Adding it to the current state array seemed to me like the simplest way to have our event processed by the upsert.